bot/exts/backend/sync/_syncers.py

```diff
@@ -161,17 +163,32 @@ async def _get_diff(self, guild: Guild) -> _Diff:
 
         for db_user in db_users.values():
             guild_user = guild_users.get(db_user.id)
+
             if guild_user is not None:
                 if db_user != guild_user:
-                    users_to_update.add(guild_user)
+                    fields_to_none: dict = {}
+
+                    for field in _User._fields:
+                        # Set un-changed values to None except ID to speed up API PATCH method.
+                        if getattr(db_user, field) == getattr(guild_user, field) and field != "id":
+                            fields_to_none[field] = None
+
+                    new_api_user = guild_user._replace(**fields_to_none)
+                    users_to_update.add(new_api_user)
```

Comment:
It's clear that a namedtuple is no longer a sufficient data structure for updates. Setting attributes to None only to remove them later is a poor design. As a comment states, namedtuples were used to facilitate easy comparison between guild and DB users. However, since you now compare attributes individually, you no longer benefit from the `tuple != other_tuple` convenience.

I don't believe the namedtuples have any advantages outside this function, so you could return a `_Diff` containing dictionaries rather than namedtuples (for roles too, for consistency).

Reply:
I was too tempted to come up with something more efficient, so here it is

```py
users_to_create = set()
users_to_update = set()
seen_guild_users = set()

async for db_user in self._get_users():
    updated_fields = {}

    def maybe_update(db_field, guild_value):
        if db_user[db_field] != guild_value:
            updated_fields[db_field] = guild_value

    if (guild_user := guild.get_member(db_user["id"])):
        seen_guild_users.add(guild_user.id)

        maybe_update("name", guild_user.name)
        maybe_update("discriminator", int(guild_user.discriminator))
        maybe_update("in_guild", True)

        guild_roles = [role.id for role in guild_user.roles]
        if set(user_dict["roles"]) != set(guild_roles):
            updated_fields["roles"] = guild_roles
    elif db_user["in_guild"]:
        updated_fields["in_guild"] = False
    
    users_to_update.add(updated_fields)

for member in guild.members:
    if member.id not in seen_guild_users:
        new_user = dict(
            id=member.id,
            name=member.name,
            discriminator=int(member.discriminator),
            roles=[role.id for role in member.roles],
            in_guild=True            
        )
        users_to_create.add(new_user)
```

Uses way less memory (not a big deal, but still nice) and has less loops.

Reply:
Sets cannot take in dicts as they are hashable, so I could use a list and check before appending or I convert the dict to a tuple to storing it in the set().